lean4-infoview/src/infoview/interactiveCode.tsx

```diff
@@ -0,0 +1,181 @@
+import * as React from 'react'
+import { Instance as TippyInstance, Props as TippyRawProps } from 'tippy.js'
+import 'tippy.js/dist/tippy.css'
+import 'tippy.js/themes/light-border.css'
+import { default as Tippy, tippy, TippyProps } from '@tippyjs/react'
+import { v4 as uuidv4 } from 'uuid'
+
+import { RpcContext } from "./contexts"
+import { DocumentPosition } from './util'
+import { CodeToken, CodeWithInfos, InfoPopup, InfoWithCtx, InteractiveDiagnostics_infoToInteractive, TaggedText } from './rpcInterface'
+
+export interface InteractiveTextComponentProps<T> {
+  pos: DocumentPosition
+  fmt: TaggedText<T>
+}
+
+export interface InteractiveTagProps<T> extends InteractiveTextComponentProps<T> {
+  tag: T
+}
+
+export interface InteractiveTaggedTextProps<T> extends InteractiveTextComponentProps<T> {
+  InnerTagUi: (_: InteractiveTagProps<T>) => JSX.Element
+}
+
+/**
+ * Core loop to display `TaggedText` objects. Invokes `InnerTagUi` on `tag` nodes in order to support
+ * various embedded information such as `InfoTree`s and `Expr`s.
+ * */
+export function InteractiveTaggedText<T>({pos, fmt, InnerTagUi}: InteractiveTaggedTextProps<T>) {
+  if ('text' in fmt) return <>{fmt.text}</>
+  else if ('append' in fmt) return <>
+    {fmt.append.map(a => <InteractiveTaggedText key={uuidv4()} pos={pos} fmt={a} InnerTagUi={InnerTagUi} />)}
+  </>
+  else if ('tag' in fmt) return <InnerTagUi pos={pos} fmt={fmt.tag[1]} tag={fmt.tag[0]} />
+  else throw `malformed 'TaggedText': '${fmt}'`
+}
+
+/**
+ * Will only render the `content` or `render` elements if the tippy is mounted to the DOM.
+ * From https://gist.github.com/atomiks/520f4b0c7b537202a23a3059d4eec908
+ */
+const LazyTippy = React.forwardRef<HTMLElement, TippyProps>((props, ref) => {
+  const [mounted, setMounted] = React.useState(false)
+
+  const lazyPlugin = {
+    fn: () => ({
+      onMount: () => setMounted(true),
+      onHidden: () => setMounted(false),
+    }),
+  }
+
+  const computedProps: React.PropsWithChildren<TippyProps> = {...props}
+
+  computedProps.plugins = [lazyPlugin, ...(props.plugins || [])]
+
+  if (props.render) {
+    computedProps.render = (...args) => (mounted && props.render ? props.render(...args) : undefined)
+  } else {
+    computedProps.content = mounted ? props.content : undefined
+  }
+
+  return <Tippy {...computedProps} ref={ref} />
+})
+
+/** Shows `explicitValue : itsType` and a docstring if there is one. */
+function TypePopupContents({pos, info, redrawTooltip}: {pos: DocumentPosition, info: InfoWithCtx, redrawTooltip: () => void}) {
+  const rs = React.useContext(RpcContext)
+  const [ip, setIp] = React.useState<InfoPopup>()
+
+  React.useEffect(() => {
+    InteractiveDiagnostics_infoToInteractive(rs, pos, info).then(val => {
+      if (val) {
+        setIp(val)
+        // We let Tippy.js know that the tooltip should be re-rendered,
```

Comment:
Yes, it's because tippy has code to auto-show and auto-hide popups on hover and keep them open when the user is hovering over the popup itself, while popper only renders the popups. Unfortunately tippy isn't quite customizable enough to do exactly what I want for the UI so there is some extra code doing that, but doing it all ourselves would still be quite a bit more complicated.

Reply:
Okay, web development is messy after all.